### PR TITLE
fix(store): warn if the zone is not actual "NgZone"

### PR DESCRIPTION
**Pull request by arturovt**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngxs/store/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information
Nooped zone is used by people who prefer to run change detection manually in their apps, this requires permanent calls to tick. If the zone is nooped then nothing will happen actually and the user will not benefit from using our default execution strategy, because the zone does nothing basically.

### CODE REVIEW

File: packages/store/tests/execution/dispatch-outside-zone-ngxs-execution-strategy.spec.ts [hunk 1]
```diff
@@ -295,62 +285,89 @@ describe('DispatchOutsideZoneNgxsExecutionStrategy', () => {
...
+        bootstrap: [MockComponent]
+      })
+      class MockModule {}
+
+      // Act
+      const warnings: string[] = [];
+
+      console.warn = (...args: string[]) => {
+        warnings.push(args[0]);
+      };
```
Comment (markwhitfeld) [inline packages/store/tests/execution/dispatch-outside-zone-ngxs-execution-strategy.spec.ts]: This should be restored back to the normal console.warn afterwards, otherwise this test could cause unexpected results.
This can be done in a `finally`.

Comment (markwhitfeld) [reply packages/store/tests/execution/dispatch-outside-zone-ngxs-execution-strategy.spec.ts]: Potentially you could make a helper method similar to `freshPlatform` called `preserveConsole` that just stores a copy of all the properties of console to another object (`const backupOfConsole = { ...console };`) then it executes the test, and then in a `finally` it writes all those properties back to console (`Object.assign(console, backupOfConsole)`);
... or something like that.

**Comment by markwhitfeld** [comment]
Just one question. If someone has nooped the ngZone then what is their intention?
How would it impact a normal app?
Can we maybe change the default ngxs execution strategy accordingly?
Just thinking out aloud.

**Comment by arturovt** [comment]
> Just one question. If someone has nooped the ngZone then what is their intention?
> How would it impact a normal app?
> Can we maybe change the default ngxs execution strategy accordingly?
> Just thinking out aloud.

Nooped zone is used by people who prefer to run change detection manually in their apps, this requires permanent calls to `tick`. If the zone is nooped then nothing will happen actually and the user will not benefit from using our default execution strategy, because the zone does nothing basically.

**Comment by arturovt** [comment]
> Potentially you could make a helper method similar to freshPlatform called preserveConsole that just stores a copy of all the properties of console to another object (const backupOfConsole = { ...console };)

I guess it could be quite complicated so I just used native Jest's `spyOn`.

File: packages/store/tests/execution/dispatch-outside-zone-ngxs-execution-strategy.spec.ts [hunk 2]
```diff
@@ -311,16 +311,13 @@ describe('DispatchOutsideZoneNgxsExecutionStrategy', () => {
...
-        warnings.push(args[0]);
-      };
+      const spy = jest.spyOn(console, 'warn').mockImplementation();
 
       await platformBrowserDynamic().bootstrapModule(MockModule, { ngZone: 'noop' });
 
       // Assert
-      expect(warnings).toEqual([CONFIG_MESSAGES[VALIDATION_CODE.ZONE_WARNING]()]);
+      expect(spy).toHaveBeenCalledWith(CONFIG_MESSAGES[VALIDATION_CODE.ZONE_WARNING]());
+      spy.mockRestore();
```
Comment (markwhitfeld) [inline packages/store/tests/execution/dispatch-outside-zone-ngxs-execution-strategy.spec.ts]: This line won't get called if the expectation on the previous line fails.

**Comment by markwhitfeld** [comment]
> Nooped zone is used by people who prefer to run change detection manually in their apps, this requires permanent calls to `tick`. If the zone is nooped then nothing will happen actually and the user will not benefit from using our default execution strategy, because the zone does nothing basically.

Please add this to the PR description...
Also, maybe the error message could recommend the `NoopNgxsExecutionStrategy`

File: packages/store/tests/execution/dispatch-outside-zone-ngxs-execution-strategy.spec.ts (hunk 2 above)
Comment (arturovt) [reply packages/store/tests/execution/dispatch-outside-zone-ngxs-execution-strategy.spec.ts]: I'm not able to call `mockRestore` before `expect` as Jest releases underneath resources.

Comment (markwhitfeld) [reply packages/store/tests/execution/dispatch-outside-zone-ngxs-execution-strategy.spec.ts]: This is why I suggested wrapping it in a function or as a finally part of a try catch block.

Comment (arturovt) [reply packages/store/tests/execution/dispatch-outside-zone-ngxs-execution-strategy.spec.ts]: Oh, okay :wink:

File: packages/store/tests/config-validator.spec.ts [hunk 3]
```diff
@@ -9,9 +9,6 @@ import { CONFIG_MESSAGES, VALIDATION_CODE } from '../src/configs/messages.config
 describe('ConfigValidator', () => {
   let validator: ConfigValidator;
   let host: HostEnvironment;
-  let actualWarning: string;
-
-  console.warn = (...args: string[]) => (actualWarning = args[0]);
```
Comment (arturovt) [inline packages/store/tests/config-validator.spec.ts]: Had to edit this file because it patches `console.warn`.

File: packages/store/tests/execution/dispatch-outside-zone-ngxs-execution-strategy.spec.ts (hunk 2 above)
Comment (arturovt) [reply packages/store/tests/execution/dispatch-outside-zone-ngxs-execution-strategy.spec.ts]: Done

File: packages/store/src/configs/messages.config.ts [hunk 4]
```diff
@@ -21,8 +21,8 @@ export const CONFIG_MESSAGES: ObjectKeyMap<Function> = {
     `State name '${current}' from ${newName} already exists in ${oldName}`,
   [VALIDATION_CODE.STATE_DECORATOR]: () => 'States must be decorated with @State() decorator',
   [VALIDATION_CODE.INCORRECT_PRODUCTION]: () =>
-    'Angular is running in production mode but NGXS is still running in the development mode!\n' +
-    'Please set developmentMode to false on the NgxsModule options when in production mode.\n' +
+    'Angular is running in the production mode but NGXS is still running in the development mode!\n' +
+    'Please set developmentMode to false on the NgxsModule options in the production mode.\n' +
```
Comment (markwhitfeld) [inline packages/store/src/configs/messages.config.ts]: Revert the changes to both lines, it was correct before.

Comment (markwhitfeld) [reply packages/store/src/configs/messages.config.ts]: Oh, and `in the development mode` should rather be `in development mode`

Comment (arturovt) [reply packages/store/src/configs/messages.config.ts]: Not a native speaker and cannot argue, just saw such statement in Angular's code:
image
But no problem for sure!

Comment (arturovt) [reply packages/store/src/configs/messages.config.ts]: Done!
